src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs

```diff
@@ -2,35 +2,105 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable
+using System.Collections.Generic;
+
 namespace System.Security.Cryptography.Encoding.Tests.Cbor
 {
-    internal ref partial struct CborValueReader
+    internal enum CborReaderState
+    {
+        UnsignedInteger = 0,
+        NegativeInteger,
+        ByteString,
+        TextString,
+        StartArray,
+        StartMap,
+        EndArray,
+        EndMap,
+        Tag,
+        Special,
+        Finished,
+        EOF
+    }
+
+    internal partial class CborReader
     {
-        private ReadOnlySpan<byte> _buffer;
+        private ReadOnlyMemory<byte> _buffer;
 
-        public CborValueReader(ReadOnlySpan<byte> buffer)
+        // remaining number of data items in current cbor context
+        // with null representing indefinite length data items.
+        // The root context ony permits one data item to be read.
+        private ulong? _remainingDataItems = 1;
+        private Stack<(CborMajorType type, ulong? remainingDataItems)>? _nestedDataItemStack;
+
+        internal CborReader(ReadOnlyMemory<byte> buffer)
         {
             _buffer = buffer;
         }
 
-        public CborInitialByte Peek()
+        public CborReaderState Peek()
         {
+            if (_remainingDataItems is null)
+            {
+                throw new NotImplementedException("indefinite length collections");
+            }
+
+            if (_remainingDataItems == 0)
+            {
+                if (_nestedDataItemStack?.Count > 0)
+                {
+                    return _nestedDataItemStack.Peek().type switch
+                    {
+                        CborMajorType.Array => CborReaderState.EndArray,
+                        CborMajorType.Map => CborReaderState.EndMap,
+                        _ => throw new Exception("CborReader internal error. Invalid CBOR major type pushed to stack."),
```

Comment:
Long term this is probably a Debug.Fail with a literal message and a default-message fallback exception of some sort.  (Since no one should see it it wouldn't get a message in the resx; but we don't hard-code exception messages... hence the compromise).